web/pandas/pdeps/0005-no-default-index-mode.md

```diff
@@ -0,0 +1,189 @@
+
+This option would not be the default. Users would need to explicitly opt-in to it, via ``pd.set_option('mode.no_default_index', True)``, via ``pd.option_context``, or via the ``PANDAS_NO_DEFAULT_INDEX`` environment variable.
+
+## Motivation and Scope
+
+The Index can be a source of confusion and frustration for pandas users. For example, let's consider the inputs
+
+```python
+In [37]: ser1 = df.groupby('sender')['amount'].sum()
+
+In [38]: ser2 = df.groupby('receiver')['amount'].sum()
+
+In [39]: ser1
+Out[39]:
+sender
+1    10
+2    15
+3    20
+5    25
+Name: amount, dtype: int64
+
+In [40]: ser2
+Out[40]:
+receiver
+1    10
+2    15
+3    20
+4    25
+Name: amount, dtype: int64
+```
+. Then:
+
+- it can be unexpected that summing `Series` with the same length (but different indices) produces `NaN`s in the result (https://stackoverflow.com/q/66094702/4451315):
+
+  ```python
+  In [41]: ser1 + ser2
+  Out[41]:
+  1    20.0
+  2    30.0
+  3    40.0
+  4     NaN
+  5     NaN
+  Name: amount, dtype: float64
+  ```
+
+- concatenation, even with `ignore_index=True`, still aligns on the index (https://github.com/pandas-dev/pandas/issues/25349):
+
+    ```python
+    In [42]: pd.concat([ser1, ser2], axis=1, ignore_index=True)
+    Out[42]:
+          0     1
+    1  10.0  10.0
+    2  15.0  15.0
+    3  20.0  20.0
+    5  25.0   NaN
+    4   NaN  25.0
+    ```
+
+- it can be frustrating to have to repeatedly call `.reset_index()` (https://twitter.com/chowthedog/status/1559946277315641345):
+
+    ```python
+    In [45]: df.value_counts(['sender', 'receiver']).reset_index().rename(columns={0: 'count'})
+    Out[45]:
+       sender  receiver  count
+    0       1         1      1
+    1       2         2      1
+    2       3         3      1
+    3       5         4      1
+    ```
+
+With this option enabled, users who don't want to worry about indices wouldn't need to.
+
+## Detailed Description
+
+This would require 3 steps:
+1. creation of a ``NoIndex`` object, which would be a subclass of ``RangeIndex`` on which
+  some operations such as ``append`` would behave differently.
+  The ``default_index`` function would then return ``NoIndex`` (rather than ``RangeIndex``) if this mode is enabled;
+2. adjusting ``DataFrameFormatter`` and ``SeriesFormatter`` to not print row labels for objects with a ``NoIndex``;
+3. adjusting methods which currently return an index to just insert a new column instead.
+
+Let's expand on all three below.
+
+### 1. NoIndex object
+
+Most of the logic could be handled within the ``NoIndex`` object.
+It would be like a ``RangeIndex``, but with the following differences:
+- `name` could only be `None`;
+- `start` could only be `0`, `step` `1`;
+- when appending an extra element, the new `Index` would still be `NoIndex`;
+- when slicing, one would still get a `NoIndex`;
```

Comment:
So if you did `df.loc[1:3, ["a", "b"]]` on a `DataFrame` with `NoIndex`, would we just ignore the first part of the slice (it's meaningless), or would we raise saying that you can't slice on a `NoIndex` ?

Reply:
I think we'd disallow `.loc` entirely with NoIndex

Reply:
`.loc` would still be allowed with boolean masks, but it'd be disallowed with labels

i.e.:
```python
In [6]: pd.set_option('mode.no_default_index', True)

In [7]: ser = pd.Series([1, 2, 3])

In [8]: ser.loc[ser>1]
Out[8]: 
2
3
dtype: int64

In [9]: ser.loc[0]
---------------------------------------------------------------------------

IndexError: Cannot use label-based indexing on NoIndex!
```

Reply:
> .loc would still be allowed with boolean masks, but it'd be disallowed with labels

what can this case handle that `.iloc` can't?

Reply:
I don't think `.iloc` handles that
```python
In [4]: ser.iloc[ser>1]
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In [4], line 1
----> 1 ser.iloc[ser>1]

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/core/indexing.py:1073, in _LocationIndexer.__getitem__(self, key)
   1070 axis = self.axis or 0
   1072 maybe_callable = com.apply_if_callable(key, self.obj)
-> 1073 return self._getitem_axis(maybe_callable, axis=axis)

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/core/indexing.py:1611, in _iLocIndexer._getitem_axis(self, key, axis)
   1608     key = np.asarray(key)
   1610 if com.is_bool_indexer(key):
-> 1611     self._validate_key(key, axis)
   1612     return self._getbool_axis(key, axis=axis)
   1614 # a list of integers

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/core/indexing.py:1452, in _iLocIndexer._validate_key(self, key, axis)
   1450 if hasattr(key, "index") and isinstance(key.index, Index):
   1451     if key.index.inferred_type == "integer":
-> 1452         raise NotImplementedError(
   1453             "iLocation based boolean "
   1454             "indexing on an integer type "
   1455             "is not available"
   1456         )
   1457     raise ValueError(
   1458         "iLocation based boolean indexing cannot use "
   1459         "an indexable as a mask"
   1460     )
   1461 return

NotImplementedError: iLocation based boolean indexing on an integer type is not available

```

Reply:
makes sense, thanks

Reply:
We could easily change `iloc` to allow this (I always have found it confusing that it doesn't), but I think the more important argument is that it is quite common to combine a boolean filter on the rows with a label-based column selection, like  `df.loc[mask, cols]`, and so this use case would then keep working (assuming the mask object also has no index). There is also no ambiguity here (masks are not actually "label" based, I would say)